mutable_tree.go

```diff
@@ -0,0 +1,469 @@
+			return nil, nil, nil, node.value, []*Node{node}
+		}
+		return node.hash, node, nil, nil, nil
+	}
+
+	// node.key < key; we go to the left to find the key:
+	if bytes.Compare(key, node.key) < 0 {
+		newLeftHash, newLeftNode, newKey, value, orphaned := tree.recursiveRemove(node.getLeftNode(tree.ImmutableTree), key)
+
+		if len(orphaned) == 0 {
+			return node.hash, node, nil, value, orphaned
+		} else if newLeftHash == nil && newLeftNode == nil { // left node held value, was removed
+			return node.rightHash, node.rightNode, node.key, value, orphaned
+		}
+		orphaned = append(orphaned, node)
+
+		newNode := node.clone(version)
+		newNode.leftHash, newNode.leftNode = newLeftHash, newLeftNode
+		newNode.calcHeightAndSize(tree.ImmutableTree)
+		newNode, balanceOrphaned := tree.balance(newNode)
+
+		return newNode.hash, newNode, newKey, value, append(orphaned, balanceOrphaned...)
+	}
+	// node.key >= key; either found or look to the right:
+	newRightHash, newRightNode, newKey, value, orphaned := tree.recursiveRemove(node.getRightNode(tree.ImmutableTree), key)
+
+	if len(orphaned) == 0 {
+		return node.hash, node, nil, value, orphaned
+	} else if newRightHash == nil && newRightNode == nil { // right node held value, was removed
+		return node.leftHash, node.leftNode, nil, value, orphaned
+	}
+	orphaned = append(orphaned, node)
+
+	newNode := node.clone(version)
+	newNode.rightHash, newNode.rightNode = newRightHash, newRightNode
+	if newKey != nil {
+		newNode.key = newKey
+	}
+	newNode.calcHeightAndSize(tree.ImmutableTree)
+	newNode, balanceOrphaned := tree.balance(newNode)
+
+	return newNode.hash, newNode, nil, value, append(orphaned, balanceOrphaned...)
+}
+
+// Load the latest versioned tree from disk.
+func (tree *MutableTree) Load() (int64, error) {
+	return tree.LoadVersion(int64(0))
+}
+
+// Returns the version number of the latest version found
+func (tree *MutableTree) LoadVersion(targetVersion int64) (int64, error) {
+	roots, err := tree.ndb.getRoots()
+	if err != nil {
+		return 0, err
+	}
+	if len(roots) == 0 {
+		return 0, nil
+	}
+	latestVersion := int64(0)
+	var latestRoot []byte
+	for version, r := range roots {
+		tree.versions[version] = true
+		if version > latestVersion &&
+			(targetVersion == 0 || version <= targetVersion) {
+			latestVersion = version
+			latestRoot = r
+		}
+	}
+
+	if !(targetVersion == 0 || latestVersion == targetVersion) {
+		return latestVersion, fmt.Errorf("wanted to load target %v but only found up to %v",
+			targetVersion, latestVersion)
+	}
+
+	t := &ImmutableTree{
+		ndb:     tree.ndb,
+		version: latestVersion,
+	}
+	if len(latestRoot) != 0 {
+		t.root = tree.ndb.GetNode(latestRoot)
+	}
+
+	tree.orphans = map[string]int64{}
+	tree.ImmutableTree = t
+	tree.lastSaved = t.clone()
+	return latestVersion, nil
+}
+
+// GetImmutable loads an ImmutableTree at a given version for querying
+func (tree *MutableTree) GetImmutable(version int64) (*ImmutableTree, error) {
+	rootHash := tree.ndb.getRoot(version)
```

Comment:
And the leaks are gone, and we get a sensible global cache behaviour of the node cache